core/modules/map.rs

```diff
@@ -1617,3 +1670,187 @@ pub fn module_origin<'a>(
     true,
   )
 }
+
+fn render_js_wasm_module(specifier: &str, wasm_deps: WasmDeps) -> String {
+  // NOTE(bartlomieju): it's unlikely the generated file will have more lines,
+  // but it's better to overallocate than to have to mem copy.
+  let mut src = Vec::with_capacity(512);
+
+  fn aggregate_wasm_module_imports(
+    imports: &[wasm_dep_analyzer::Import],
+  ) -> IndexMap<String, Vec<String>> {
+    let mut imports_map = IndexMap::default();
+
+    for import in imports.iter().filter(|i| {
+      matches!(i.import_type, wasm_dep_analyzer::ImportType::Function(..))
+    }) {
+      let entry = imports_map
+        .entry(import.module.to_string())
+        .or_insert(vec![]);
+      entry.push(import.name.to_string());
+    }
+
+    imports_map
+  }
+
+  src.push(format!(
+    r#"import wasmMod from "{}" with {{ type: "$$deno-core-internal-wasm-module" }};"#,
+    specifier,
+  ));
+
+  // TODO(bartlomieju): handle imports collisions?
+  if !wasm_deps.imports.is_empty() {
+    let aggregated_imports = aggregate_wasm_module_imports(&wasm_deps.imports);
+
+    for (key, value) in aggregated_imports.iter() {
+      src.push(format!(
+        r#"import {{ {} }} from "{}";"#,
+        value.join(", "),
+        key
+      ));
+    }
+
+    src.push("const importsObject = {".to_string());
+
+    for (key, value) in aggregated_imports.iter() {
+      src.push(format!("  \"{}\": {{", key).to_string());
+
+      for el in value {
+        src.push(format!("    {},", el));
+      }
+
+      src.push("  },".to_string());
+    }
+
+    src.push("};".to_string());
+
+    src.push(
+      "const modInstance = await import.meta.wasmInstantiate(wasmMod, importsObject);".to_string(),
+    )
+  } else {
+    src.push(
+      "const modInstance = await import.meta.wasmInstantiate(wasmMod);"
+        .to_string(),
+    )
+  }
+
+  if !wasm_deps.exports.is_empty() {
+    for export_desc in wasm_deps.exports.iter().filter(|e| {
+      matches!(e.export_type, wasm_dep_analyzer::ExportType::Function)
+    }) {
+      if export_desc.name == "default" {
+        src.push(format!(
+          "export default modInstance.exports.{};",
+          export_desc.name
+        ));
+      } else {
+        src.push(format!(
+          "export const {} = modInstance.exports.{};",
+          export_desc.name, export_desc.name
+        ));
+      }
+    }
+  }
+
+  src.join("\n")
+}
+
+#[test]
+fn test_render_js_wasm_module() {
+  let deps = WasmDeps {
+    imports: vec![],
+    exports: vec![],
+  };
+  let rendered = render_js_wasm_module("./foo.wasm", deps);
+  pretty_assertions::assert_eq!(
+    rendered,
+    r#"import wasmMod from "./foo.wasm" with { type: "$$deno-core-internal-wasm-module" };
+const modInstance = await import.meta.wasmInstantiate(wasmMod);"#,
+  );
+
+  let deps = WasmDeps {
```

Comment:
We could probably store some basic WASM modules in the testdata and simplify these tests.

Reply:
`WasmDeps` doesn't implement `Serialize`/`Deserialize` so there's no easy way to store intermediate representation. I agree that it could be improved, but maybe it's not a blocker for this PR?